include/babeltrace/durin.h

```diff
@@ -0,0 +1,218 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <dwarf.h>
+#include <elfutils/libdw.h>
+#include <babeltrace/babeltrace-internal.h>
+
+/*
+ * Durin is a wrapper over libdw providing a nicer, higher-level
+ * interface, to access basic debug information. Durin is also the
+ * eldest of the seven Fathers of the Dwarves created by the Vala Aulë
+ * in the lore of Tolkien's Middle-earth.
+ */
+
+/*
+ * This structure corresponds to a single compilation unit (CU) for a
+ * given set of debug information (Dwarf type).
+ */
+struct durin_cu {
+	Dwarf *dwarf_info;
+	/* Offset in bytes in the DWARF file to current CU header. */
+	Dwarf_Off offset;
+	/* Offset in bytes in the DWARF file to next CU header. */
+	Dwarf_Off next_offset;
+	/* Size in bytes of CU header */
+	size_t header_size;
+};
+
+/*
+ * This structure represents a single debug information entry (DIE),
+ * within a compilation unit (CU).
+ */
+struct durin_die {
+	struct durin_cu *cu;
+	Dwarf_Die *dwarf_die;
+	/*
+	 * A depth of 0 represents a root DIE, located in the DWARF
+	 * layout on the same level as its corresponding CU entry. Its
+	 * children DIEs will have a depth of 1, and so forth.
+	 */
+	unsigned int depth;
+};
+
+/**
+ * Instantiate a structure to access compile units (CU) from a given
+ * `dwarf_info`.
+ *
+ * @param dwarf_info	Dwarf instance
+ * @returns		Pointer to the new durin_cu on success,
+ *			NULL on failure.
+ */
+BT_HIDDEN
+struct durin_cu *durin_cu_create(Dwarf *dwarf_info);
+
+/**
+ * Destroy the given durin_cu instance.
+ *
+ * @param cu	durin_cu instance
+ */
+BT_HIDDEN
+void durin_cu_destroy(struct durin_cu *cu);
+
+/**
+ * Advance the compile unit `cu` to the next one.
+ *
+ * On success, `cu`'s offset is set to that of the current compile
+ * unit in the executable. On failure, `cu` remains unchanged.
+ *
+ * @param cu	durin_cu instance
+ * @returns	0 on success, -1 on failure
+ */
+BT_HIDDEN
+int durin_cu_next(struct durin_cu *cu);
+
+/**
+ * Instantiate a structure to access debug information entries (DIE)
+ * for the given compile unit `cu`.
+ *
+ * @param cu	durin_cu instance
+ * @returns	Pointer to the new durin_die on success,
+ *		NULL on failure.
+ */
+BT_HIDDEN
+struct durin_die *durin_die_create(struct durin_cu *cu);
+
+/**
+ * Destroy the given durin_die instance.
+ *
+ * @param die	durin_die instance
+ */
+BT_HIDDEN
+void durin_die_destroy(struct durin_die *die);
+
+/**
+ * Advance the debug information entry `die` to its first child, if
+ * any.
+ *
+ * @param die	durin_die instance
+ * @returns	0 on success, -1 on failure
+ */
+BT_HIDDEN
+int durin_die_child(struct durin_die *die);
```

Comment:
This function (as well as `durin_die_next` and others) should distinguish between a hard error (i.e. failed allocation) and no child DIE available by returning -1 for the former and 1 for the latter.